File: server/src/main/java/io/microconfig/server/configs/DynamicVarsResolverStrategy.java
package io.microconfig.server.configs;

import io.microconfig.core.properties.DeclaringComponentImpl;
import io.microconfig.core.properties.Placeholder;
import io.microconfig.core.properties.PlaceholderResolveStrategy;
import io.microconfig.core.properties.Property;
import lombok.RequiredArgsConstructor;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static io.microconfig.core.properties.ConfigFormat.PROPERTIES;
import static io.microconfig.core.properties.PropertyImpl.property;
import static java.util.Optional.empty;
import static java.util.Optional.of;

@RequiredArgsConstructor
public class DynamicVarsResolverStrategy implements PlaceholderResolveStrategy {
    private final Map<String, String> vars;

    public Map<String, String> dynamicVars() {
        return new HashMap<>(vars);
    }

    @Override
    public Optional<Property> resolve(Placeholder placeholder) {
        var value = vars.get(placeholder.getKey());
        if (value == null) return empty();

        return of(property(placeholder.getKey(), value, PROPERTIES,
                new DeclaringComponentImpl(placeholder.getConfigType(), "Dynamic Vars", placeholder.getEnvironment())));
    }

}

File: server/src/main/kotlin/io/microconfig/server/configs/DynamicVarsResolverStrategy.kt
package io.microconfig.server.configs

import io.microconfig.core.properties.ConfigFormat.PROPERTIES
import io.microconfig.core.properties.DeclaringComponentImpl
import io.microconfig.core.properties.PlaceholderResolveStrategy
import io.microconfig.core.properties.Property
import io.microconfig.core.properties.PropertyImpl.property
import io.microconfig.server.common.toOptional
import java.util.Optional
import java.util.Optional.empty

class DynamicVarsResolverStrategy(private val vars: Map<String, String>) : PlaceholderResolveStrategy {

    override fun resolve(
        component: String,
        key: String,
        environment: String,
        configType: String,
        root: String
    ): Optional<Property> {
        val value = vars[key] ?: return empty()
        val parent = DeclaringComponentImpl(configType, "Dynamic Vars", environment)
        return property(key, value, PROPERTIES, parent).toOptional()
    }
}